### PR TITLE
[Feature] Support encryption mode

### PR DESCRIPTION
**Pull request by zhangheihei**
## Why I'm doing:
Currently the aes_encrypt/aes_dencrypt functions only support AES_128_ECB.
#### 1.1 ECB Mode (Electronic Codebook)
- `AES_128_ECB`, `AES_192_ECB`, `AES_256_ECB`
- No IV required
- Not recommended for production (no IV means identical plaintext blocks produce identical ciphertext)

#### 1.2 CBC Mode (Cipher Block Chaining)
- `AES_128_CBC`, `AES_192_CBC`, `AES_256_CBC`
- Requires 16-byte IV
- Most commonly used mode

#### 1.3 CFB Mode (Cipher Feedback)
- `AES_128_CFB`, `AES_192_CFB`, `AES_256_CFB`
- `AES_128_CFB1`, `AES_192_CFB1`, `AES_256_CFB1`
- `AES_128_CFB8`, `AES_192_CFB8`, `AES_256_CFB8`
- `AES_128_CFB128`, `AES_192_CFB128`, `AES_256_CFB128`
- Requires 16-byte IV
- Stream cipher mode

#### 1.4 OFB Mode (Output Feedback)
- `AES_128_OFB`, `AES_192_OFB`, `AES_256_OFB`
- Requires 16-byte IV
- Stream cipher mode

#### 1.5 CTR Mode (Counter)
- `AES_128_CTR`, `AES_192_CTR`, `AES_256_CTR`
- Requires 16-byte IV
- Stream cipher mode

#### 1.6 GCM Mode (Galois/Counter Mode)
- `AES_128_GCM`, `AES_192_GCM`, `AES_256_GCM`
- Requires 12-byte IV
- Supports AAD (Additional Authenticated Data)
- Provides both encryption and authentication
- Output format: [IV][Ciphertext][TAG]

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

---

> [!NOTE]
> Add AES encrypt/decrypt support for ECB/CBC/CFB/OFB/CTR/GCM (with IV and optional AAD), keeping 2-param forms compatible via FE, with full BE impl, tests, and docs.
> 
> - **Backend (BE)**:
>   - **AES Util**: Extend `AesUtil` with `encrypt_ex`/`decrypt_ex`, support modes `ECB/CBC/CFB(1/8/128)/OFB/CTR/GCM`, IV handling (defaults), and GCM AAD/tag; add helpers `get_mode_from_string`, `is_gcm_mode|is_stream_mode|is_ecb_mode`.
>   - **Encryption Functions**: Replace `aes_encrypt`/`aes_decrypt` with `aes_encrypt_with_mode`/`aes_decrypt_with_mode`; add param extraction (mode/key/iv/aad), NULL handling, reusable buffers; keep 2-param variants (default `AES_128_ECB`).
> - **Frontend (FE)**:
>   - **ExpressionAnalyzer**: Auto-convert 2/3 params to 4 (`data,key[,iv],mode='AES_128_ECB'`); validate 5th AAD only with GCM.
>   - **FunctionSet**: Treat `AES_ENCRYPT`/`AES_DECRYPT` as not-always-null-with-null-param.
> - **Function Registration**:
>   - Register overloads for `aes_encrypt`/`aes_decrypt` (2/4/5 params) pointing to new BE fns.
> - **Tests**:
>   - Add extensive BE unit tests for all modes (ECB/CBC/CFB/OFB/CTR/GCM with/without AAD), NULL cases, constants, batch/perf; update AES util tests to new API.
>   - Add SQL tests covering new signatures and validations.
> - **Docs**:
>   - Update `aes_encrypt`/`aes_decrypt` syntax to include `[iv][, mode][, aad]`, list supported modes, and add examples (EN/JA/ZH).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e45f6f04895580b4cc35e81df2e9271520328763. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by wangsimo0** [comment]
thanks for contributing, amazing! could you pls add docs?

**Comment by zhangheihei** [comment]
> thanks for contributing, amazing! could you pls add docs?

I'll add doc later

**Comment by alvin-phoenix-ai** [comment]
@cursor review

**Comment by alvin-phoenix-ai** [comment]
@cursor review

File: be/src/exprs/encryption_functions.cpp [hunk 1]
```diff
@@ -25,67 +27,280 @@
...
 
+        // Calculate output buffer size for decryption
+        // Decrypted plaintext will never exceed ciphertext size
         int cipher_len = src_value.size;
-        char p[cipher_len];
 
-        int len = AesUtil::decrypt(AES_128_ECB, (unsigned char*)src_value.data, src_value.size,
-                                   (unsigned char*)key_value.data, key_value.size, nullptr, true, (unsigned char*)p);
+        // Use stack allocation (VLA - Variable Length Array)
+        char decrypt_buf[cipher_len];
```
[the remark this answers is not in the document]
Comment (zhangheihei) [reply be/src/exprs/encryption_functions.cpp]: fix VLA

**Review by Seaven — DISMISSED** [review]

**Review by trueeyu — DISMISSED** [review]

File: gensrc/script/functions.py [hunk 2]
```diff
@@ -884,8 +884,16 @@
     [110112, "json_contains", False, False, "BOOLEAN", ["JSON", "JSON"], "JsonFunctions::json_contains"],
 
     # aes and base64 function
-    [120100, "aes_encrypt", False, False, "VARCHAR", ["VARCHAR", "VARCHAR"], "EncryptionFunctions::aes_encrypt"],
-    [120110, "aes_decrypt", False, False, "VARCHAR", ["VARCHAR", "VARCHAR"], "EncryptionFunctions::aes_decrypt"],
+        # aes_encrypt: 4-parameter version (data, key, iv, mode)
+    # Note: FE's ExpressionAnalyzer converts 2/3 params to 4 params automatically
+    [120101, "aes_encrypt", False, False, "VARCHAR", ["VARCHAR", "VARCHAR", "VARCHAR", "VARCHAR"], "EncryptionFunctions::aes_encrypt_with_mode"],
```
Comment (trueeyu) [inline gensrc/script/functions.py:891]: If be upgraded first, there may be compilation problems?

**Review by trueeyu — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Comment by kevincai** [comment]
As marked as "Feature", this will be only available in next Release (v4.1), will not backport to existing releases.

**Comment by alvin-phoenix-ai** [comment]
@cursor review

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -1032,8 +1036,64 @@ public Void visitFunctionCall(FunctionCallExpr node, Scope scope) {
...
+                    // Only GCM mode supports AAD parameter
+                    if (node.getChildren().size() == 5) {
+                        Expr modeExpr = node.getChild(3);
+                        if (modeExpr instanceof StringLiteral) {
+                            String mode = ((StringLiteral) modeExpr).getValue().toUpperCase();
+                            if (!mode.contains("GCM")) {
+                                throw new SemanticException(
+                                        fnName + " with 5 parameters requires GCM mode to use AAD parameter, " +
+                                                "but got mode: " + mode + ". " +
+                                                "Only GCM modes (AES_128_GCM, AES_192_GCM, AES_256_GCM) support AAD parameter.",
```
[the remark this answers is not in the document]
Comment (zhangheihei) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:1089]: For GCM MODE, the FE does not require strict behavior checks; it only provides users with reasonable usage prompts. The BE side performs strict behavior checks.

**Review by dirtysalt — APPROVED** [review]
